Crux Data Browser - Entity State

danmason opened the pull request:
Adds `_entity`, as per #788

danmason committed:
Add _entity and use muuntaja for content negotiation

danmason committed:
Add support for valid-time and transaction-time query params

danmason committed:
Add 404 on entity-id returning nil

danmason committed:
Change transaction-time -> transact-time

jarohen commented on crux-http-server/src/crux/http_server.clj:
```diff
@@ -393,10 +412,40 @@
      (->HTTPServer server options))))
 
+(defn- edn->html [edn]
+  (format
```
what will this do with nested structures?

jarohen commented on crux-http-server/src/crux/http_server.clj:
```diff
@@ -393,10 +412,40 @@
   {::server {:start-fn (fn [{:keys [crux.node/node]} {::keys [port] :as options}]
                          (let [server (j/run-jetty (-> (partial handler node)
                                                        (p/wrap-params)
+                                                       (wrap-format
```
this will wrap all the endpoints with this HTML formatter - that what we want?

jarohen commented on crux-http-server/src/crux/http_server.clj:
```diff
@@ -199,6 +203,18 @@
         (add-last-modified tx-time))))
 
+(defn- entity-state [^ICruxAPI crux-node request]
+  (let [[_ encoded-eid] (re-find #"^/_entity/(.+)$" (req/path-info request))]
```
I'd say the time's come to start with a proper routing library - we've outgrown manually parsing out each individual value

jarohen replied:
as discussed, bringing this in as a separate change

jarohen commented on crux-http-server/src/crux/http_server.clj:
```diff
@@ -199,6 +203,18 @@
+                                                            (instant/read-instant-date))
+                                        :transact-time (some-> (get query-params "transaction-time")
+                                                               (instant/read-instant-date))})
+          entity-map (.entity db eid)]
```
advise using the Clojure API here, `(crux/entity db eid)`

jarohen commented on crux-http-server/src/crux/http_server.clj:
```diff
@@ -393,10 +412,40 @@
+       ^String (edn->html data)
+       ^String charset))
+    mfc/EncodeToOutputStream
+    (encode-to-output-stream [_ data charset]
```
the reason for a separate `encode-to-output-stream` (in addition to `encode-to-bytes`) is likely that the former's to be used for large streaming response bodies (which we will likely need for queries), the latter being sufficient for when the response fits in memory.

probably more of an issue when you implement `/_query`, so maybe this is fine for now

jarohen replied:
yep, let's leave this for query

jarohen reviewed (changes requested):
thanks Dan - few things to think about below. 

also, can't see a build for this commit?

danmason committed:
Move out /_entity/ to its' own handler

Allows it to only apply the 'wrap-format' middleware on that one
particular route

danmason committed:
Make entity->html return <ul> elements, handle nested structures

danmason committed:
Use hiccup, and use dls for maps and uls for lists

danmason committed:
Use api/entity

jarohen committed:
indentation/formatting

jarohen committed:
double newline

jarohen committed:
ensuring we covered all the cases in edn->html
